go/pkg/services/log_collector/server.go

```diff
@@ -16,36 +16,507 @@ package log_collector
+
+	// create a log file to the pod
+	filePath := lcs.resolvePodLogFilePath(runId, podName)
+	if err := common.EnsureFileExists(filePath); err != nil {
+		lcs.Logger.ErrorWithCtx(ctx,
+			"Failed to ensure log file",
+			"runId", runId,
+			"filePath", filePath)
+		return
+	}
+
+	// get logs from pod, and keep the stream open (follow)
+	podLogOptions := &v1.PodLogOptions{
+		Follow: true,
+	}
+	restClientRequest := lcs.kubeClientSet.CoreV1().Pods(lcs.namespace).GetLogs(podName, podLogOptions)
+
+	// initialize error for the while loop
+	streamErr := errors.New("Stream error")
+	var stream io.ReadCloser
+
+	// stream logs - retry if failed
+	for streamErr != nil {
+		stream, streamErr = restClientRequest.Stream(ctx)
+		if streamErr != nil {
+			lcs.Logger.WarnWithCtx(ctx, "Failed to get pod log read/closer", "runId", runId)
+		}
+		time.Sleep(1 * time.Second)
+	}
+	defer stream.Close() // nolint: errcheck
+
+	for {
+		buf := make([]byte, 1024)
+
+		// This is blocking until there is something to read
+		numBytes, err := stream.Read(buf)
+
+		// nothing read, continue
+		if numBytes == 0 {
+			continue
+		}
+
+		// if error is EOF, the pod is done streaming logs (deleted/completed/failed)
+		if err == io.EOF {
+			lcs.Logger.DebugWithCtx(ctx, "Pod logs are done streaming", "runId", runId)
+			break
+		}
+
+		// if error is not EOF, log it and continue
+		if err != nil {
+			lcs.Logger.WarnWithCtx(ctx, "Failed to read pod log",
+				"error",
+				err, "runId", runId)
+			continue
+		}
+
+		// write log contents to file
+		if err := common.WriteToFile(ctx, lcs.Logger, filePath, buf[:numBytes], true); err != nil {
+			lcs.Logger.ErrorWithCtx(ctx,
+				"Failed to write log contents to file",
+				"runId", runId,
+				"podName", podName)
+		}
+	}
+
+	lcs.Logger.DebugWithCtx(ctx,
+		"Removing item from state file",
+		"runId", runId,
+		"podName", podName)
+
+	// remove run id from state file
+	if err := lcs.removeItemFromInProgress(runId); err != nil {
+		lcs.Logger.WarnWithCtx(ctx, "Failed to remove run id from state file")
+	}
+}
+
+// addItemToInProgress adds an item to the `in_progress` list in the state file
+func (lcs *LogCollectorServer) addItemToInProgress(ctx context.Context, runId string, selector string) error {
+
+	state := lcs.getState()
+
+	logItem := LogItem{
+		RunId:         runId,
+		LabelSelector: selector,
+	}
+
+	if existingItem, exists := state.InProgress[runId]; exists {
+		lcs.Logger.DebugWithCtx(ctx,
+			"Item already exists in state file. Overwriting label selector",
+			"runId", runId,
+			"existingItem", existingItem)
+		// TODO: notify the goroutines to stop and restart them with the new selector
+	}
+
+	state.InProgress[runId] = logItem
+
+	lcs.setState(state)
+
+	return nil
+}
+
+// removeItemFromInProgress removes an item from the `in_progress` list in the state file
+func (lcs *LogCollectorServer) removeItemFromInProgress(runId string) error {
```

Comment:
```suggestion
func (lcs *LogCollectorServer) removeItemFromInProgress(runId string) {
```
no errors in this method